src/dune_engine/build_system.ml

```diff
@@ -2,6 +2,10 @@ open Import
 open Memo.O
 module Action_builder = Action_builder0
 
+(* CR-someday amokhov: We should just get rid of this. It's broken in
+   file-watching mode (see [sandbox-mkdir.t]) and is also unnecessary: we should
+   just [mkdir] a directory before [chdir]ing into it because it's cheap and we
+   are already [mkdir]ing a lot -- see [Sandbox.create_dirs]. *)
```

Comment:
An easy way to fix the bug would be to depend on the current run in this function. In fact, we could then use it to create sandboxed dirs as well. IMO the optimization is small but is still worth it. We cache stat calls for digests for example.